### PR TITLE
rustbuild: Start building --enable-extended

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit adds a new flag to the configure script,
`--enable-extended`, which is intended for specifying a desire to
compile the full suite of Rust tools such as Cargo, the RLS, etc. This
is also an indication that the build system should create combined
installers such as the pkg/exe/msi artifacts.

Currently the `--enable-extended` flag just indicates that combined
installers should be built, and Cargo is itself not compiled just yet
but rather only downloaded from its location. The intention here is to
quickly get to feature parity with the current release process and then
we can start improving it afterwards.

All new files in this PR inside `src/etc/installer` are copied from the
rust-packaging repository.

cc #38531

### CODE REVIEW

**Comment by alexcrichton** [comment]
@brson ok I've updated with support for Windows as well, this should be good to go.

File: src/bootstrap/dist.rs [hunk 1]
```diff
@@ -500,3 +507,365 @@ fn write_file(path: &Path, data: &[u8]) {
...
+       cmd.env("CFG_MINGW", "0")
+          .env("CFG_ABI", "MSVC");
+    }
+
+    if target.contains("x86_64") {
+       cmd.env("CFG_PLATFORM", "x64");
+    } else {
+       cmd.env("CFG_PLATFORM", "x86");
+    }
+}
```
Comment (brson) [inline src/bootstrap/dist.rs]: This is all amazing. Thank you.

**Comment by brson** [comment]
This all looks great, but can we get rid of the 'platform' branding? How about `--enable-extended`, for 'extended build', or 'extended toolset'.

r=me

**Pull request merged**